crates/sqlexec/src/planner/session_planner.rs

```diff
@@ -959,17 +917,78 @@ impl<'a> SessionPlanner<'a> {
                     }
                 };
 
-                TableOptions::Excel(TableOptionsExcel {
+                TableOptionsExcel {
                     location,
                     storage_options,
                     file_type: None,
                     compression: None,
                     sheet_name,
                     has_header,
-                })
+                }
+                .into()
             }
             other => return Err(internal!("unsupported datasource: {}", other)),
-        };
+        })
+    }
+
+    async fn plan_create_external_table(
+        &self,
+        mut stmt: CreateExternalTableStmt,
+    ) -> Result<LogicalPlan> {
+        let datasource = normalize_ident(stmt.datasource);
+
+        let tunnel = stmt.tunnel.map(normalize_ident);
+        let tunnel_options = self.get_tunnel_opts(&tunnel)?;
+        if let Some(tunnel_options) = &tunnel_options {
+            // Validate if the tunnel type is supported by the datasource
+            validate_table_tunnel_support(&datasource, tunnel_options.as_str()).map_err(|e| {
+                PlanError::InvalidExternalTable {
+                    source: Box::new(e),
+                }
+            })?;
+        }
+
+        let creds = stmt.credentials.map(normalize_ident);
+        let creds_options = self.get_credentials_opts(&creds)?;
+        if let Some(creds_options) = &creds_options {
+            validate_table_creds_support(&datasource, creds_options.as_str()).map_err(|e| {
+                PlanError::InvalidExternalTable {
+                    source: Box::new(e),
+                }
+            })?;
+        }
+
+
+        let schema = stmt
+            .columns
+            .map(|columns| {
+                let fields = columns
+                    .into_iter()
+                    .map(|coll| -> Result<Field, PlanError> {
+                        // check if there is a NOT NULL constraint
+                        let is_nullable = !coll
+                            .options
+                            .into_iter()
+                            .any(|k| matches!(k.option, ColumnOption::NotNull));
```

Comment:
given the application of user-specified schema for external data sources is that people _probably_ never actually want an _not null_ constraint and I'd be inclined to error early when such schemas are specified and not attempt to enforce the constraint.

Reply:
i think that is an assumption we shouldn't make, but I'll make the changes regardless.